prov/gni/src/gnix_ep.c

```diff
@@ -2051,10 +2055,238 @@ DIRECT_FN int gnix_ep_open(struct fid_domain *domain, struct fi_info *info,
 		ret = gnix_nic_alloc(domain_priv, NULL, &ep_priv->nic);
 		if (ret != FI_SUCCESS) {
 			GNIX_WARN(FI_LOG_EP_CTRL,
-				    "_gnix_nic_alloc call returned %d\n",
-				     ret);
+				    "gnix_nic_alloc call returned %d\n", ret);
+			goto err;
+		}
+	}
+
+	/*
+	 * if smsg callbacks not present hook them up now
+	 */
+
+	if (ep_priv->nic->smsg_callbacks == NULL)
+		ep_priv->nic->smsg_callbacks = gnix_ep_smsg_callbacks;
+
+	_gnix_ref_get(ep_priv->domain);
+	*ep = &ep_priv->ep_fid;
+	return ret;
+
+err:
+	if (ep_priv->xpmem_hndl) {
+		if (_gnix_xpmem_handle_destroy(ep_priv->xpmem_hndl) !=
+		    FI_SUCCESS) {
+			GNIX_WARN(FI_LOG_EP_CTRL,
+				  "_gnix_xpmem_handle_destroy returned %s\n",
+				  fi_strerror(-ret));
+		}
+	}
+
+	__destruct_tag_storages(ep_priv);
+
+	if (free_list_inited == true)
+		__fr_freelist_destroy(ep_priv);
+
+	if (ep_priv->cm_nic != NULL)
+		ret = _gnix_cm_nic_free(ep_priv->cm_nic);
+
+	if (ep_priv->nic != NULL)
+		ret = _gnix_nic_free(ep_priv->nic);
+
+	free(ep_priv);
+	return ret;
+
+}
+
+int _gnix_ep_alloc(struct fid_domain *domain, struct fi_info *info,
```

Comment:
I found this name confusing since it's only used for scalable endpoints.  A lot of things are common with regular endpoints, but also a lot is not common.  Is it worth factoring gnix_ep_open to use this?  If not, I think the name should be changed to be something like `_gnix_sep_alloc`.